### ISSUE TITLE
change projection matrix to camera intrinsics

### ISSUE THREAD

**Issue by CanCanZeng** [hbb1/2d-gaussian-splatting#110]
Hello, the camera model of my own data is not an ideal pinhole model, the principal point is not in the center of the image. Therefore, I want to change the **projection matrix** in the code to **camera intrinsics**. But I found that the result of forward pass is correct, but when calculating the gradient in backward pass, there will be a lot of NaN s. This problem has been bothering me for several days. If the author has time, can you take a look and roughly point out where the problem may lie?

The code address is `https://github.com/CanCanZeng/2d-gaussian-splatting` and `https://github.com/CanCanZeng/diff-surfel-rasterization` You can directly use `git clone git@github.com CanCanZeng/2d-gaussian-splitting.git 2d-gaussian-splittingzcc -- recursive` to clone the complete code to the local.

I have made minimal changes to the source code to facilitate debugging. I see that the only places where the projection matrix is used in the forward are in **`in_frustum`** and **`computete_transmat`**. However, in_frustum does not actually use the results of the projection matrix, so only computete_transmat has been modified, and the core idea remains consistent with the original code. It seems that the projection matrix is only used in the **`compute_transmat_aabb`** function in the backward, and it is used in the same way as in the forward, so the code is also the same as in the forward. I printed the **`T`** matrix in the forward and backward sections and found that there were no abnormal situations.

**Comment by hbb1** [comment]
Hi, supporting a non-ideal-pinhole model is interesting. I will review your implementation and come back to you then.

By the way, many folks and I have made some progress toward this goal. From my understanding, we should be able to support this now, without modifying the cuda code but simply adapting a new [projection matrix] (https://github.com/hbb1/2d-gaussian-splatting/pull/74)  given works have been done in https://github.com/hbb1/2d-gaussian-splatting/commit/b8c9128beec9133511cf61fbacd23efc72c562b4 and https://github.com/hbb1/2d-gaussian-splatting/commit/21a32f11449f0e787bc4b622e7e13c736086fa5b. You can also have a try.

**Comment by hbb1** [comment]
Also, ensure that your projection matrix is consistent with the original implementation when using the ideal pinhole model, which your model should be compatible with.

**Comment by CanCanZeng** [comment]
Hi @hbb1 , I just tried this commit https://github.com/hbb1/2d-gaussian-splatting/commit/b8c9128beec9133511cf61fbacd23efc72c562b4 today, but I found that the recovered intrinsics here is not consistent with the original parameters given in the colmap. You can take a break point here to check if the values of the `intrins `are completely consistent with the values given in the colmap file.

By the way, I don't intend to convince you to add my modification to your source code, because my modification will make the interface inconsistent with the original 3DGS. But the way of calculating the projection using the intrinsic matrix (**`K`**) is ​​more in line with the way of thinking of people who do computer vision, so I changed it to the form I am more accustomed to..

**Comment by hbb1** [comment]
The intrins can be recover from the projection matrix by doing so
```python
ndc2pix = torch.tensor([
        [W / 2, 0, 0, (W) / 2],
        [0, H / 2, 0, (H) / 2],
        [0, 0, 0, 1]]).float().cuda().T
intrins = (view.projection_matrix @ ndc2pix)[:3,:3].T
```
I mean you should make sure you modification should work equally for an ideal pinhole model.

**Comment by hbb1** [comment]
You can also check this version that also use intrins https://github.com/hbb1/diff-surfel-rasterization/commit/661d22484ac4fe15495d34410eab4c635b3ceb74

**Comment by CanCanZeng** [comment]
> The intrins can be recover from the projection matrix by doing so
> 
> ```python
> ndc2pix = torch.tensor([
>         [W / 2, 0, 0, (W) / 2],
>         [0, H / 2, 0, (H) / 2],
>         [0, 0, 0, 1]]).float().cuda().T
> intrins = (view.projection_matrix @ ndc2pix)[:3,:3].T
> ```
> 
> I mean you should make sure you modification should work equally for an ideal pinhole model.

Of course, both ideal and non-ideal pinhole camera models are supported.

<br>

> You can also check this version that also use intrins [hbb1/diff-surfel-rasterization@661d224](https://github.com/hbb1/diff-surfel-rasterization/commit/661d22484ac4fe15495d34410eab4c635b3ceb74)

Thank you for the reminder. I am currently reviewing this part of the code.

**Comment by CanCanZeng** [comment]
I found the bug, a very inconspicuous one. Here https://github.com/CanCanZeng/diff-surfel-rasterization/blob/a463669255a6624537afeb1db1491c9ff64c21a3/cuda_rasterizer/backward.cu#L503, I should not use a local variable **`P`**, but the **`P`** declared outside. The code prompt of vscode is too weak. The variable is used without assigning a value. If it is under other IDE, there will be a warning.

**Comment by hbb1** [comment]
Ok! A very unnoticeable bug. That makes sense.
